backend/de.metas.edi/src/main/java/de/metas/edi/api/impl/DesadvDAO.java

```diff
@@ -64,26 +66,36 @@ public class DesadvDAO implements IDesadvDAO
 	private static final String SYS_CONFIG_DefaultMinimumPercentage = "de.metas.esb.edi.DefaultMinimumPercentage";
 	private static final String SYS_CONFIG_DefaultMinimumPercentage_DEFAULT = "50";
 	private static final String SYS_CONFIG_MATCH_USING_BPARTNER_ID = "de.metas.edi.desadv.MatchUsingC_BPartner_ID";
+	private static final String SYS_CONFIG_MATCH_USING_ORDER_ID = "de.metas.edi.desadv.MatchUsingC_Order_ID";
 
 	private final IQueryBL queryBL = Services.get(IQueryBL.class);
 	private final ISysConfigBL sysConfigBL = Services.get(ISysConfigBL.class);
 
 	@Override
-	public I_EDI_Desadv retrieveMatchingDesadvOrNull(@NonNull final String poReference, @NonNull final BPartnerId bPartnerId, @NonNull final IContextAware ctxAware)
+	public I_EDI_Desadv retrieveMatchingDesadvOrNull(@NonNull final EDIDesadvQuery query)
 	{
-		Check.assumeNotEmpty(poReference, "Param 'poReference' is not emtpy; ctxAware={}", ctxAware);
-		final IQueryBuilder<I_EDI_Desadv> query = queryBL.createQueryBuilder(I_EDI_Desadv.class, ctxAware)
+		final IQueryBuilder<I_EDI_Desadv> queryBuilder = queryBL.createQueryBuilder(I_EDI_Desadv.class, query.getCtxAware())
 				.addOnlyActiveRecordsFilter()
-				.addEqualsFilter(I_EDI_Desadv.COLUMN_POReference, poReference)
+				.addEqualsFilter(I_EDI_Desadv.COLUMN_POReference, query.getPoReference())
 				.addEqualsFilter(I_EDI_Desadv.COLUMN_Processed, false)
 				.addEqualsFilter(I_EDI_Desadv.COLUMN_Processing, false);
 
 		if (isMatchUsingBPartnerId())
 		{
-			query.addEqualsFilter(I_EDI_Desadv.COLUMNNAME_C_BPartner_ID, bPartnerId.getRepoId());
+			queryBuilder.addEqualsFilter(I_EDI_Desadv.COLUMNNAME_C_BPartner_ID, query.getBPartnerRepoId());
 		}
 
-		return query.create().firstOnly(I_EDI_Desadv.class);
+		if (isMatchUsingOrderId(query.getCtxAware()) && query.getOrderId() != null)
```

Comment:
IMHO in DAO u shall not check the sysconfig,
you shall simply filter by orderId if u got it in the query.

It's the job of the caller to properly set query orderId or not based on sysconfigs or whatever.